GitUI/UserControls/NativeTreeView.cs

```diff
@@ -8,4 +12,100 @@ protected override void CreateHandle()
+
+        // Invoked just before an inner tree node is expanded/collapsed. Set CancelEventHandler.Cancel to true
+        // to cancel the expand/collapse of the current node.
+        public event CancelEventHandler BeforeDoubleClickExpandCollapse;
+
+        public NativeTreeViewDoubleClickDecorator(NativeTreeView treeView)
+        {
+            _treeView = treeView;
+            _treeView.Disposed += OnDisposed;
+            _treeView.BeforeCollapse += OnBeforeCollapse;
+            _treeView.BeforeExpand += OnBeforeExpand;
+            _treeView.MouseDown += OnMouseDown;
+        }
+
+        private void OnDisposed(object sender, EventArgs e)
+        {
+            _treeView.BeforeCollapse -= OnBeforeCollapse;
+            _treeView.BeforeExpand -= OnBeforeExpand;
+            _treeView.MouseDown -= OnMouseDown;
+            _treeView = null;
+        }
+
+        private void OnBeforeCollapse(object sender, TreeViewCancelEventArgs e)
+        {
+            HandleBeforeDoubleClickExpandCollapse(sender, e);
+        }
+
+        private void OnBeforeExpand(object sender, TreeViewCancelEventArgs e)
+        {
+            HandleBeforeDoubleClickExpandCollapse(sender, e);
+        }
+
+        private void HandleBeforeDoubleClickExpandCollapse(object sender, TreeViewCancelEventArgs e)
+        {
+            // Once detected, reset double-clicked state so that if BeforeDoubleClickExpandCollapse does
+            // anything that collapses/expands nodes on the tree, this function won't erroneously resend
+            // this event.
+            bool doubleClicked = DoubleClicked();
+            ResetDoubleClick();
+
+            if (doubleClicked)
+            {
+                var cancelEventArgs = new CancelEventArgs();
+                BeforeDoubleClickExpandCollapse?.Invoke(sender, cancelEventArgs);
+                e.Cancel = cancelEventArgs.Cancel;
+            }
+        }
+
+        private void OnMouseDown(object sender, MouseEventArgs e)
+        {
+            // We only care about double-clicks on the node itself, not the plus/minus part
+            var hitTest = _treeView.HitTest(e.Location);
+            if (hitTest.Location == TreeViewHitTestLocations.PlusMinus)
+            {
+                return;
+            }
+
+            // Use _mouseClicks and _lastMouseDown to detect double-clicks, and set _lastDoubleClick when detected
+            ++_mouseClicks;
+
+            if (_mouseClicks == 2)
+            {
+                int delta = (int)DateTime.Now.Subtract(_lastMouseDown).TotalMilliseconds;
+                if (delta >= 0 && delta < SystemInformation.DoubleClickTime)
+                {
+                    _lastDoubleClick = DateTime.Now;
+                }
+
+                _mouseClicks = 0;
+            }
+
+            _lastMouseDown = DateTime.Now;
+        }
+
+        // Returns true if a double-click occured very recently
+        private bool DoubleClicked()
```

Comment:
I'm not finding this method's name particularly meaningful... And I struggle to suggest a good name either.
😞

Reply:
"WasDoubleClicked"?

Reply:
`IsDoubleClickStateSet` 👍